OpenRA.Mods.Common/FileFormats/InstallShieldCABCompression.cs

```diff
@@ -0,0 +1,438 @@
+		{
+			this.volumes = volumes;
+
+			if (header.ReadUInt32() != 0x28635349)
+				throw new InvalidDataException("Not an Installshield CAB package");
+
+			header.Position += 8;
+			var cabDescriptorOffset = header.ReadUInt32();
+			header.Position = cabDescriptorOffset + 12;
+			var cabDescriptor = new CabDescriptor(header);
+			header.Position += 14;
+
+			var fileGroupOffsets = new uint[MaxFileGroupCount];
+			for (var i = 0; i < MaxFileGroupCount; i++)
+				fileGroupOffsets[i] = header.ReadUInt32();
+
+			header.Position = cabDescriptorOffset + cabDescriptor.FileTableOffset;
+			var directories = new DirectoryDescriptor[cabDescriptor.DirectoryCount];
+			for (var i = 0; i < directories.Length; i++)
+				directories[i] = new DirectoryDescriptor(header, cabDescriptorOffset + cabDescriptor.FileTableOffset);
+
+			var fileGroups = new List<FileGroup>();
+			foreach (var offset in fileGroupOffsets)
+			{
+				var nextOffset = offset;
+				while (nextOffset != 0)
+				{
+					header.Position = cabDescriptorOffset + (long)nextOffset + 4;
+					var descriptorOffset = header.ReadUInt32();
+					nextOffset = header.ReadUInt32();
+					header.Position = cabDescriptorOffset + descriptorOffset;
+
+					fileGroups.Add(new FileGroup(header, cabDescriptorOffset));
+				}
+			}
+
+			header.Position = cabDescriptorOffset + cabDescriptor.FileTableOffset + cabDescriptor.FileTableOffset2;
+			foreach (var fileGroup in fileGroups)
+			{
+				for (var i = fileGroup.FirstFile; i <= fileGroup.LastFile; i++)
+				{
+					header.Position = cabDescriptorOffset +	cabDescriptor.FileTableOffset + cabDescriptor.FileTableOffset2 + i * 0x57;
+					var file = new FileDescriptor(header, i, cabDescriptorOffset + cabDescriptor.FileTableOffset);
+					var path = "{0}\\{1}\\{2}".F(fileGroup.Name, directories[file.DirectoryIndex].Name, file.Filename);
+					index[path] = file;
+				}
+			}
+		}
+
+		public void ExtractFile(string filename, Stream output, Action<int> onProgress = null)
+		{
+			FileDescriptor file;
+			if (!index.TryGetValue(filename, out file))
+				throw new FileNotFoundException(filename);
+
+			ExtractFile(file, output, onProgress);
+		}
+
+		void ExtractFile(FileDescriptor file, Stream output, Action<int> onProgress = null)
+		{
+			if (file.Flags.HasFlag(CABFlags.FileInvalid))
+				throw new Exception("File Invalid");
+
+			if (file.LinkFlags.HasFlag(LinkFlags.Prev))
+			{
+				var prev = index.Values.First(f => f.Index == file.LinkToPrevious);
+				ExtractFile(prev, output, onProgress);
+				return;
+			}
+
+			if (file.Flags.HasFlag(CABFlags.FileObfuscated))
+				throw new NotImplementedException("Obfuscated files are not supported");
+
+			var extracter = new CabExtracter(file, volumes);
+			extracter.CopyTo(output, onProgress);
+
+			if (output.Length != file.ExpandedSize)
+				throw new Exception("Did not fully extract Expected = {0}, Got = {1}".F(file.ExpandedSize, output.Length));
```

Comment:
`InvalidDataException` maybe? and `Did not fully extract.` with a sentence end might make this easier to read.